[Xelians] Fix vulnerability cas server

pyXelians opened the pull request:
Cette PR permet de corriger des vulnérabilités de sécurités présentes dans cas server. 

## Type de changement

-  Code et/ou Configuration : Correction

## Tests

Suite au build un rapport de test est généré avec les vulnérabilités. 

## Migration

 Quels sont les opérations supplémentaires à faire lors d’une migration de VITAM ? 
 De quel version à quel version s’appliquent ces modifications ?

## Checklist

 [ * ] Les tests unitaires nouveaux et existants passent avec succès localement.
 [ * ] Toutes les dépendances ont été mergées en priorité

## Contributeur

Xelians

leleuj commented on cas/cas-server/src/main/java/fr/gouv/vitamui/cas/webflow/actions/DispatcherAction.java:
```diff
@@ -141,6 +141,7 @@ protected Event doExecute(final RequestContext requestContext) throws IOExceptio
         }
         val response = WebUtils.getHttpServletResponseFromExternalWebflowContext(requestContext);
         val webContext = new JEEContext(request, response);
+        sessionStore.set(webContext, Constants.PROVIDER_TECHNICAL_NAME, provider.getTechnicalName());
```
Pourquoi cet ajout en session ?

Cela me semble un vieux bout de code maintenant inutile car le provider technical name est pris des `ClientCredential` (dans le `UserPrincipalResolver`).

leleuj commented on cas/cas-server/src/main/java/fr/gouv/vitamui/cas/webflow/actions/CustomSendTokenAction.java:
```diff
@@ -110,7 +112,7 @@ protected Event doExecute(final RequestContext requestContext) {
         // add the obfuscated phone to the webflow in case of success
         if (CasWebflowConstants.TRANSITION_ID_SUCCESS.equals(event.getId())) {
-            requestContext.getFlowScope().put("mobile", obfuscateMobile(mobile));
+            requestContext.getFlowScope().put("mobile", escapeHtml4(obfuscateMobile(mobile)));
```
OK

leleuj replied:
J'ai peur que ça fasse double emploi avec : https://github.com/ProgrammeVitam/vitam-ui/pull/646/files#diff-0aa54ced11f5684bd687eee43f51f988d79f0dc3d7281e21433afa9eeb5f5db9R18 et https://github.com/ProgrammeVitam/vitam-ui/pull/646/files#diff-f733ab0f604a4b646387ed00fcd6f2f9ed7e5f87530911e5d474e3b9797eff31R57

Un numéro de mobile n'ayant pas de caractère HTML, ça ne sera sûrement pas visible néanmoins comme problème...

leleuj commented on cas/cas-server/src/main/resources/templates/casPwdView.html:
```diff
@@ -29,7 +29,7 @@ <h5 th:if="${!#strings.isEmpty(superUser)}">
                 </h5>
 
-                <h5 th:if="${#strings.isEmpty(superUser)}" th:utext="#{cas.authentication.welcomeback.msg (${credential.username})}">
+                <h5 th:if="${#strings.isEmpty(superUser)}" th:text="#{cas.authentication.welcomeback.msg (${credential.username})}">
```
OK

leleuj commented on cas/cas-server/src/main/resources/templates/casSmsCodeExpiredView.html:
```diff
@@ -10,12 +10,12 @@
             <input type="hidden" name="execution" th:value="${flowExecutionKey}"/>
             <input type="hidden" name="_eventId" value="resend"/>
-            <a class="button" th:utext="#{screen.mfa.sms.expired.resend(${mobile})}" href="javascript:void(0)" onclick="$('#resendCodeForm').submit();"/>
+            <a class="button" th:text="#{screen.mfa.sms.expired.resend(${mobile})}" href="javascript:void(0)" onclick="$('#resendCodeForm').submit();"/>
```
OK

leleuj commented on cas/cas-server/src/main/resources/templates/casSmsMissingPhoneView.html:
```diff
@@ -10,11 +10,11 @@
+        <h4 th:text="#{screen.mfa.sms.missingphone.title}"></h4>
 
-        <div class="text large" th:utext="#{screen.mfa.sms.missingphone.text1(${firstname})}"></div>
+        <div class="text large" th:text="#{screen.mfa.sms.missingphone.text1(${firstname})}"></div>
```
OK

leleuj commented:
L'idée initiale est bonne et même nécessaire: protéger les données non sûres en les HTML-escapant. C'est là où j'ai mis "Ok".
Pour le reste, HTML-escaper tous les textes va se révéler contre productif pour les messages définis avec un HTML.

**Exemples:**
```
screen.logout.success=Vous vous êtes déconnecté(e). Vous pouvez vous <a href="login"><span class="text large">connecter</span></a> à nouveau.
```

et

```
screen.mfa.sms.login.resetcode1=Vous n&apos;avez pas reçu de code par SMS ?
```

ne vont plus s'afficher correctement...

pyXelians commented:
> 

Je comprends ta réponse, bien qu'elle n'apporte pas forcément de solution. Pour les deux exemples que tu m'as donné nous avons fait une "correction" rapide, pour éviter d'avoir à retravailler tous les messages.

leleuj commented:
Pour moi, la solution est de ne HTML-encoder que les champs qui viennent d'une saisie utilisateur ou de la base : username, mobile, firstname...
Les messages définis en dur dans les properties ne doivent pas être HTML encodés (comportement CAS par défaut).

pyXelians commented:
Ok, pour le coup les modifications que l'on a faites vont dans ce sens. Après de manière générale pour éviter des copiés collés malheureux (sans réflexion) on a préféré escape tous les  champs, pour que le fait d'interprété du HTML devienne une exception et pas le comportement par défaut. Mais on aura toujours la possibilité si on le désire d'ajouter des exceptions qui seront réfléchis. Cela te va ?

leleuj commented:
Personnellement et dans l'esprit de CAS, j'aurais préférer ne HTML-encoder que les textes à problème.
Après, ce n'est pas un gros souci de trop HTML-encoder. Donc oui, ça me va sur la partie HTML-encoding des messages.

Sur la constante `PROVIDER_TECHNICAL_NAME`, elle ne sert à rien, il faut l'enlever de la PR.

pyXelians commented:
la constante est utilisée dans le fichier cas/cas-server/src/main/java/fr/gouv/vitamui/cas/webflow/actions/DispatcherAction.java

leleuj commented:
Oui, mais cette variable stockée dans le `sessionStore` n'est jamais relue. C'est du code que j'avais écrit et changé. A virer donc.

pyXelians commented:
> Oui, mais cette variable stockée dans le `sessionStore` n'est jamais relue. C'est du code que j'avais écrit et changé. A virer donc.

La constante a été supprimée.

leleuj reviewed (approved):
C'est bon pour moi maintenant

leleuj commented:
@bbenaissa tu viens de forcer un push et la constante `PROVIDER_TECHNICAL_NAME` est revenue alors qu'elle avait été supprimée. Il y a un problème !

bbenaissa commented:
Hello @leleuj, mon action c'était simplemet un rebase à partir de develop, si la constante existe, c'est qu'elle n'a pas été supprimée

leleuj commented:
Bizarre, bizarre, il me semble bien qu'elle a été supprimée dans un commit précédent : @pyXelians ?

bbenaissa commented:
@leleuj Je ne sais pas ce que c'est passé, je viens de remettre le commit en question

bbenaissa commented:
Pour le besoin de cherry-picker cette PR, je sqash les commits en un seul en reprenant l'ensemble des messages de commit

bbenaissa committed:
[TECH-815] Fix html encoding for cas
[CAS] Use SAML attribute for profile id instead of SAML id
[TECH] Fix CAS security issue
